File: hs/app/Parser.hs
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE TupleSections #-}

module Parser where

import qualified Ast
import qualified BasicBlock
import Control.Applicative (Alternative (empty), (<|>))
import Control.Monad ((>=>))
import Data.Bifunctor (second)
import Data.List (find)
import qualified Function
import qualified Insn
import Lexer
import qualified Op
import qualified Type
import qualified Value

type Stream = [Token]

newtype Parser a = Parser {parseFun :: Stream -> Maybe (Stream, a)}

instance Functor Parser where
  fmap :: (a -> b) -> Parser a -> Parser b
  -- Help: (>=>) :: Monad m => (a -> m b) -> (b -> m c) -> a -> m c
  fmap f (Parser g) = Parser (g >=> (return . second f))

instance Applicative Parser where
  pure :: a -> Parser a
  -- Construct a parser that does not consume string, only return a value.
  pure x = Parser (\s -> return (s, x))

  (<*>) :: Parser (a -> b) -> Parser a -> Parser b
  Parser f <*> g =
    Parser (f >=> (\(s, cons) -> parseFun (fmap cons g) s))

instance Alternative Parser where
  empty :: Parser a
  empty = Parser (const Nothing)

  (<|>) :: Parser a -> Parser a -> Parser a
  f <|> g = Parser (\s -> parseFun f s <|> parseFun g s)

-- Construct a parser that extracts a first token.
take1 :: Parser Token
take1 =
  Parser
    ( \case
        (x : xs) -> return (xs, x)
        _ -> Nothing
    )

-- Construct a parser that check if an input starts from
-- one of the enumerated tokens.
oneOf :: [Token] -> Parser Token
oneOf l =
  Parser
    (parseFun take1 >=> (\(rest, token) -> fmap (rest,) (find (token ==) l)))

-- Construct a parser that replaces other parse error with Nothing.
optional :: Parser a -> Parser (Maybe a)
optional p =
  Parser
    ( \s -> case parseFun p s of
        Just (rest, e) -> return (rest, Just e)
        Nothing -> return (s, Nothing)
    )

-- Parse a type.
typ :: Parser Type.Type
typ =
  (cons <$> fmap f (oneOf [Keyword Void, Keyword Unsigned, Keyword Int, Keyword Bool]))
    <*> fold0 (\x _ -> x + 1) 0 (tok Star)
  where
    f :: Token -> Type.Type
    f = \case
      Keyword Void -> Type.Void
      Keyword Unsigned -> Type.Unsigned32
      Keyword Int -> Type.Int32
      Keyword Bool -> Type.Bool
      _ -> error "unreachable"

    -- TODO: not build list here.
    cons :: Type.Type -> Int -> Type.Type
    cons tp num = foldr (\_ t -> Type.Ref t) tp [0 .. num - 1]

-- Parse specified token.
tok :: Token -> Parser Token
tok t = oneOf [t]

-- Parse any ident.
ident :: Parser Token
ident =
  Parser
    ( parseFun take1
        >=> ( \(rest, token) -> case token of
                Ident _ -> return (rest, token)
                _ -> Nothing
            )
    )

-- Parse a number.
number :: Parser Int
number =
  Parser
    ( parseFun take1
        >=> ( \(rest, token) -> case token of
                Number n -> return (rest, n)
                _ -> Nothing
            )
    )

-- Parse a literal.
literal :: Parser Token
literal =
  (cons <$> optional (oneOf [Plus, Minus])) <*> number
  where
    cons sg n =
      case sg of
        (Just Plus) -> Number n
        Nothing -> Number n
        Just Minus -> Number (-n)
        _ -> error "unrechable"

-- Parse a separated with delimeter `delim` list of `p`.
separated_list0 :: Parser a -> Parser e -> Parser [e]
separated_list0 delim p =
  Parser
    ( \s -> case parseFun p s of
        Just (stail, e0) -> return $ second (e0 :) (h stail)
        Nothing -> return (s, [])
    )
  where
    h s = case parseFun delim s of
      Nothing -> (s, [])
      Just (no_delim, _) -> case parseFun p no_delim of
        Nothing -> (s, [])
        Just (rest, e) -> second (e :) (h rest)

-- Parse a pair.
pair :: Parser a -> Parser b -> Parser (a, b)
pair f g = ((,) <$> f) <*> g

-- Parse a function signature.
sign :: Parser (Type.Type, String, [(Type.Type, Value.Variable)])
sign =
  (cons <$> typ)
    <*> fmap fromIdent ident
    <*> ( tok LPar
            *> separated_list0
              (tok Comma)
              ( pair
                  typ
                  (fmap ((\n -> Value.Variable {Value.name = n}) . fromIdent) ident)
              )
            <* tok RPar
        )
  where
    cons foo_typ name args = (foo_typ, name, args)

-- Parse a value.
value :: Parser Value.Value
value = cons <$> (ident <|> literal)
  where
    cons t = case t of
      Ident n -> Value.ValueVariable {Value.var = Value.Variable {Value.name = n}}
      Number n -> Value.ValueLiteral {Value.lit = Value.Literal {Value.val = n}}
      _ -> error "unreachable"

-- Parse a return instruction.
ret :: Parser Insn.Insn
ret =
  cons <$> (tok (Keyword Return) *> value)
  where
    cons x = Insn.Return {Insn.ret = x}

-- Parse a binary operation.
binaryOp :: Parser (Value.Value, Op.BinaryOp, Value.Value)
binaryOp = (cons <$> value) <*> operator <*> value
  where
    operator :: Parser Op.BinaryOp
    operator =
      opCons
        <$> oneOf
          [ Plus,
            Minus,
            Star,
            Slash,
            Percent,
            Gt,
            Lt,
            And,
            Xor,
            Or,
            Eq,
            Gq,
            Lq,
            Shl,
            Shr
          ]
      where
        opCons token = case token of
          Plus -> Op.Arithm Op.Add
          Minus -> Op.Arithm Op.Sub
          Star -> Op.Arithm Op.Mul
          Slash -> Op.Arithm Op.Div
          Percent -> Op.Arithm Op.Mod
          Gt -> Op.Logical Op.Gt
          Lt -> Op.Logical Op.Lt
          And -> Op.Logical Op.And
          Xor -> Op.Logical Op.Xor
          Or -> Op.Logical Op.Or
          Eq -> Op.Logical Op.Eq
          Gq -> Op.Logical Op.Gq
          Lq -> Op.Logical Op.Lq
          Shl -> Op.Bit Op.Shl
          Shr -> Op.Bit Op.Shr
          _ -> error "unreachable"

    cons lhs op rhs = (lhs, op, rhs)

-- Parse an unary operation.
unaryOp :: Parser (Op.UnaryOp, Value.Value)
unaryOp = (cons <$> operator) <*> value
  where
    operator :: Parser Op.UnaryOp
    operator =
      opCons
        <$> oneOf
          [Minus, Plus, Star, Not, And, Tilda]
      where
        opCons token = case token of
          Minus -> Op.Minus
          Plus -> Op.Plus
          Not -> Op.Not
          Star -> Op.Deref
          And -> Op.RefOf
          Tilda -> Op.Neg
          _ -> error "unreachable"

    cons op arg = (op, arg)

-- Parse a function call.
call :: Parser Value.FCall
call =
  (cons <$> ident)
    <*> (tok LPar *> separated_list0 (tok Comma) value <* tok RPar)
  where
    cons name args = Value.FCall {Value.fun = fromIdent name, Value.args = args}

-- Parse a rvalue.
rvalue :: Parser Value.RValue
rvalue =
  (toUnaryOp <$> unaryOp)
    <|> ( toBinaryOp
            <$> binaryOp
        )
    -- Call must be considered prior to value, because of:
    -- `int a = b();` vs `int a = b;`
    <|> (Value.Call <$> call)
    <|> (Value.Val <$> value)
  where
    toUnaryOp (op, arg) = Value.Unary {Value.uop = op, Value.arg = arg}
    toBinaryOp (lhs, op, rhs) = Value.Binary {Value.lhs = lhs, Value.bop = op, Value.rhs = rhs}

variable :: Parser Value.Variable
variable = (\n -> Value.Variable {Value.name = fromIdent n}) <$> ident

-- Parse an assign statement.
assign :: Parser Insn.Insn
assign = (cons <$> variable) <*> (tok Assign *> rvalue)
  where
    cons var rhs = Insn.Assign {Insn.lhs = var, Insn.rhs = rhs}

-- Parse an assign deref statement.
assignDeref :: Parser Insn.Insn
assignDeref = (cons <$> (tok Star *> variable <* tok Assign)) <*> value
  where
    cons var rhs = Insn.AssignDeref {Insn.lhs = var, Insn.drhs = rhs}

-- Parse a goto instruction.
gotoInsn :: Parser Insn.GotoInsn
gotoInsn = cons <$> (tok (Keyword Goto) *> ident)
  where
    cons n = Insn.GotoInsn {Insn.label = fromIdent n}

-- Parse a goto instruction as a generic instruction.
goto :: Parser Insn.Insn
goto = fmap Insn.Goto gotoInsn

-- Parse a conditional operator.
iff :: Parser Insn.Insn
iff =
  (cons <$> (tok (Keyword If) *> tok LPar *> value <* tok RPar))
    <*> gotoInsn
    <*> optional (tok Semicolon *> tok (Keyword Else) *> gotoInsn)
  where
    cons condition thn oth =
      Insn.If {Insn.condition = condition, Insn.thn = thn, Insn.oth = oth}

-- Parse a declaration.
decl :: Parser Insn.Insn
decl = (cons <$> typ) <*> variable <*> optional (tok Assign *> rvalue)
  where
    cons tp var rhs = Insn.Declaration {Insn.typ = tp, Insn.lhs = var, Insn.declRhs = rhs}

-- Parse an instruction.
insn :: Parser Insn.Insn
insn = ret <|> assign <|> assignDeref <|> goto <|> iff <|> decl

-- TODO(@askalt): think about laziness vs energy here.
fold0 :: (b -> a -> b) -> b -> Parser a -> Parser b
fold0 f ini p = Parser (Just . h ini)
  where
    h acc s = case parseFun p s of
      Nothing -> (s, acc)
      Just (suff, e) ->
        h (f acc e) suff

-- Apply parser until it not fail and collect results to the list.
many0 :: Parser a -> Parser [a]
many0 p = reverse <$> fold0 (\xs x -> x : xs) [] p

-- Parse a basic block.
basicBlock :: Parser BasicBlock.BasicBlock
basicBlock = (cons <$> ident) <*> (tok Colon *> many0 (pair insn (tok Semicolon)))
  where
    cons n insns =
      BasicBlock.BasicBlock
        { BasicBlock.label = fromIdent n,
          BasicBlock.insns = map fst insns
        }

-- Parse a function body.
body :: Parser [BasicBlock.BasicBlock]
body = many0 basicBlock

-- Parse a function.
function :: Parser Function.Function
function = (cons <$> sign) <*> (tok LCb *> body <* tok RCb)
  where
    cons (tp, name, args) b =
      Function.Function
        { Function.return_type = tp,
          Function.name = name,
          Function.args = args,
          Function.blocks = b
        }

-- Parse a module.
ast :: Parser Ast.Ast
ast = cons <$> many0 function
  where
    cons funs = Ast.Ast {Ast.funs = funs}

-- Parse a token list with specified parser.
parseWith :: Parser a -> String -> Either Error a
parseWith p s = do
  tokens <- lex0 s
  case parseFun p tokens of
    Nothing -> Left $ ParseError Nothing
    Just (rest, parsed) -> case rest of
      [] -> return parsed
      _ -> Left $ ParseError (Just rest)

-- Main parse runner.
parse :: String -> Either Error Ast.Ast
parse = parseWith ast

File: parser/src/parser.rs
#![allow(clippy::type_complexity)]

use nom::{
    branch::alt,
    combinator::{map, map_res},
    error::{Error as NomError, ErrorKind, ParseError},
    multi::{fold_many0, many0, separated_list0},
    sequence::{pair, tuple},
    IResult,
};

use crate::{
    ast::{
        ArithmBinaryOp, Ast, BasicBlock, BinaryOp, BitBinaryOp, Call, Function, GotoInsn, Insn,
        Literal, LogicalBinaryOp, RValue, Type, UnaryOp, Value, Variable,
    },
    lexer::{
        Keyword, Lexer,
        Token::{self, *},
    },
};

use super::{Error, Result};

fn nom_error(i: &[Token], kind: ErrorKind) -> nom::Err<NomError<&[Token]>> {
    nom::Err::Error(NomError::from_error_kind(i, kind))
}

/// Combinator that takes one token from the input.
fn take1(i: &[Token]) -> IResult<&[Token], Token> {
    let (first_token, remaining) = i
        .split_first()
        .ok_or_else(|| nom_error(i, ErrorKind::Complete))?;
    Ok((remaining, first_token.clone()))
}

/// Combinator that checks if an input starts from one of the enumerated tokens.
fn one_of<'a>(tokens: &'a [Token]) -> impl Fn(&'a [Token]) -> IResult<&'a [Token], Token> {
    move |i| {
        map_res(take1, |first_token| {
            tokens
                .iter()
                .find(|&token| *token == first_token)
                .map(|_| first_token)
                .ok_or_else(|| nom_error(i, ErrorKind::OneOf))
        })(i)
    }
}

/// Combinator that wraps the result of the other combinator with option,
/// returning [`None`] if it return an recoverable error.
fn maybe<I: Clone, O, E: ParseError<I>, F>(
    mut parser: F,
) -> impl FnMut(I) -> IResult<I, Option<O>, E>
where
    F: nom::Parser<I, O, E>,
{
    move |i| {
        let i_cloned = i.clone();
        let res = parser.parse(i);
        match res {
            Ok((i, o)) => Ok((i, Some(o))),
            Err(nom::Err::Error(_)) => Ok((i_cloned, None)),
            Err(err) => Err(err),
        }
    }
}

/// Parse a type.
fn typ(i: &[Token]) -> IResult<&[Token], Type> {
    map(
        pair(
            one_of(&[
                Token::Keyword(Keyword::Void),
                Token::Keyword(Keyword::Int),
                Token::Keyword(Keyword::Unsigned),
                Token::Keyword(Keyword::Bool),
            ]),
            fold_many0(tok(Token::Star), || 0, |s, _| s + 1),
        ),
        |(token, stars)| {
            let typ = match token {
                Token::Keyword(Keyword::Void) => Type::Void,
                Token::Keyword(Keyword::Int) => Type::Int32,
                Token::Keyword(Keyword::Unsigned) => Type::Unsigned32,
                Token::Keyword(Keyword::Bool) => Type::Bool,
                _ => unreachable!(),
            };
            (0..stars).fold(typ, |acc, _| Type::Ref(Box::new(acc)))
        },
    )(i)
}

/// Parse a specified token.
fn tok(token: Token) -> impl Fn(&[Token]) -> IResult<&[Token], Token> {
    move |i| {
        map_res(take1, |first_token| {
            if token == first_token {
                Ok(token.clone())
            } else {
                Err(nom_error(i, ErrorKind::Tag))
            }
        })(i)
    }
}

/// Parse an ident.
fn ident(i: &[Token]) -> IResult<&[Token], Token> {
    map_res(take1, |token| match token {
        Token::Ident(_) => Ok(token),
        _ => Err(nom_error(i, ErrorKind::Tag)),
    })(i)
}

/// Parse a literal.
fn literal(i: &[Token]) -> IResult<&[Token], Token> {
    map_res(
        pair(maybe(one_of(&[Token::Plus, Token::Minus])), take1),
        |(sign, token)| match token {
            Token::Number(n) => match sign {
                Some(Token::Plus) | None => Ok(Token::Number(n)),
                // TODO: overflow.
                Some(Token::Minus) => Ok(Token::Number(-n)),
                _ => unreachable!(),
            },
            _ => Err(nom_error(i, ErrorKind::Tag)),
        },
    )(i)
}

/// Parse a function signature.
fn signature(i: &[Token]) -> IResult<&[Token], (Type, String, Vec<(Type, Variable)>)> {
    map(
        tuple((
            // Actual grammar is here.
            typ,
            ident,
            tok(LPar),
            separated_list0(tok(Comma), pair(typ, ident)),
            tok(RPar),
        )),
        |(return_type, name, _, decl, _)| {
            (
                return_type,
                name.into_ident().unwrap(),
                decl.into_iter()
                    .map(|(typ, name)| {
                        (
                            typ,
                            Variable {
                                name: name.into_ident().unwrap(),
                            },
                        )
                    })
                    .collect(),
            )
        },
    )(i)
}

/// Parse a `return` instruction.
fn ret(i: &[Token]) -> IResult<&[Token], Insn> {
    map(
        pair(tok(Token::Keyword(Keyword::Return)), value),
        |(_, value)| Insn::Return(value),
    )(i)
}

/// Parse a variable.
fn variable(i: &[Token]) -> IResult<&[Token], Variable> {
    map(ident, |token| Variable {
        name: token.into_ident().unwrap(),
    })(i)
}

/// Parse a binary operation.
fn binary_op(i: &[Token]) -> IResult<&[Token], (Value, BinaryOp, Value)> {
    fn operator(i: &[Token]) -> IResult<&[Token], BinaryOp> {
        map(
            one_of(&[
                Token::Plus,
                Token::Minus,
                Token::Star,
                Token::Slash,
                Token::Percent,
                Token::Gt,
                Token::Lt,
                Token::And,
                Token::Xor,
                Token::Or,
                Token::Eq,
                Token::Gq,
                Token::Lq,
                Token::Shl,
                Token::Shr,
            ]),
            |token| match token {
                // Arithmetic.
                Token::Plus => BinaryOp::ArithmOp(ArithmBinaryOp::Add),
                Token::Minus => BinaryOp::ArithmOp(ArithmBinaryOp::Sub),
                Token::Star => BinaryOp::ArithmOp(ArithmBinaryOp::Mul),
                Token::Slash => BinaryOp::ArithmOp(ArithmBinaryOp::Div),
                Token::Percent => BinaryOp::ArithmOp(ArithmBinaryOp::Mod),

                // Logical.
                Token::Gt => BinaryOp::LogicalOp(LogicalBinaryOp::Gt),
                Token::Lt => BinaryOp::LogicalOp(LogicalBinaryOp::Lt),
                Token::And => BinaryOp::LogicalOp(LogicalBinaryOp::And),
                Token::Xor => BinaryOp::LogicalOp(LogicalBinaryOp::Xor),
                Token::Or => BinaryOp::LogicalOp(LogicalBinaryOp::Or),
                Token::Eq => BinaryOp::LogicalOp(LogicalBinaryOp::Eq),
                Token::Gq => BinaryOp::LogicalOp(LogicalBinaryOp::Gq),
                Token::Lq => BinaryOp::LogicalOp(LogicalBinaryOp::Lq),

                Token::Shl => BinaryOp::BitOp(BitBinaryOp::Shl),
                Token::Shr => BinaryOp::BitOp(BitBinaryOp::Shr),
                _ => unreachable!(),
            },
        )(i)
    }

    tuple((value, operator, value))(i)
}

/// Parse an unary operation.
fn unary_op(i: &[Token]) -> IResult<&[Token], (UnaryOp, Value)> {
    fn operator(i: &[Token]) -> IResult<&[Token], UnaryOp> {
        map(
            one_of(&[
                Token::Minus,
                Token::Plus,
                Token::Star,
                Token::Not,
                Token::And,
                Token::Tilda,
            ]),
            |token| match token {
                Token::Minus => UnaryOp::Minus,
                Token::Plus => UnaryOp::Plus,
                Token::Not => UnaryOp::Not,
                Token::Star => UnaryOp::Deref,
                Token::And => UnaryOp::Ref,
                Token::Tilda => UnaryOp::Neg,
                _ => unreachable!(),
            },
        )(i)
    }

    pair(operator, value)(i)
}

/// Parse a function call.
fn call(i: &[Token]) -> IResult<&[Token], Call> {
    map(
        tuple((
            ident,
            tok(Token::LPar),
            separated_list0(tok(Token::Comma), value),
            tok(Token::RPar),
        )),
        |(fun, _, args, _)| Call {
            fun: fun.into_ident().unwrap(),
            args,
        },
    )(i)
}

/// Parse a rvalue.
fn rvalue(i: &[Token]) -> IResult<&[Token], RValue> {
    alt((
        map(unary_op, |(op, arg)| RValue::UnaryOp { op, arg }),
        map(binary_op, |(lhs, op, rhs)| RValue::BinaryOp {
            lhs,
            op,
            rhs,
        }),
        // Call must be considered prior to value, because of:
        // `int a = b();` vs `int a = b;`
        map(call, RValue::Call),
        map(value, RValue::Value),
    ))(i)
}

/// Parse a value.
fn value(i: &[Token]) -> IResult<&[Token], Value> {
    map(alt((ident, literal)), |token| match token {
        Token::Ident(n) => Value::Variable(Variable { name: n }),
        Token::Number(n) => Value::Literal(Literal { val: n }),
        _ => unreachable!(),
    })(i)
}

/// Parse an assign statement.
fn assign(i: &[Token]) -> IResult<&[Token], Insn> {
    map(
        tuple((variable, tok(Token::Assign), rvalue)),
        |(lhs, _, rhs)| Insn::Assign { lhs, rhs },
    )(i)
}

/// Parse an assign dereference statement.
fn assign_deref(i: &[Token]) -> IResult<&[Token], Insn> {
    map(
        tuple((tok(Token::Star), variable, tok(Token::Assign), value)),
        |(_, lhs, _, rhs)| Insn::AssignDeref { lhs, rhs },
    )(i)
}

/// Parse a goto instruction.
fn goto(i: &[Token]) -> IResult<&[Token], Insn> {
    map(
        pair(tok(Token::Keyword(Keyword::Goto)), ident),
        |(_, label)| {
            Insn::Goto(GotoInsn {
                label: label.into_ident().unwrap(),
            })
        },
    )(i)
}

/// Parse a conditional operator.
fn iff(i: &[Token]) -> IResult<&[Token], Insn> {
    map(
        tuple((
            tok(Token::Keyword(Keyword::If)),
            tok(Token::LPar),
            value,
            tok(Token::RPar),
            goto,
            maybe(tuple((
                tok(Token::Semicolon),
                tok(Token::Keyword(Keyword::Else)),
                goto,
            ))),
        )),
        |(_, _, condition, _, then, otherwise)| Insn::If {
            condition,
            then: then.into_goto().unwrap(),
            otherwise: otherwise.map(|(_, _, goto)| goto.into_goto().unwrap()),
        },
    )(i)
}

/// Parse a variable declaration.
fn declaration(i: &[Token]) -> IResult<&[Token], Insn> {
    map(
        tuple((typ, variable, maybe(pair(tok(Token::Assign), rvalue)))),
        |(typ, lhs, rhs)| Insn::Declaration {
            typ,
            lhs,
            rhs: rhs.map(|rhs| rhs.1),
        },
    )(i)
}

/// Parse a instruction.
fn insn(i: &[Token]) -> IResult<&[Token], Insn> {
    alt((ret, assign, assign_deref, goto, iff, declaration))(i)
}

/// Parse a basic block.
fn basic_block(i: &[Token]) -> IResult<&[Token], BasicBlock> {
    map(
        tuple((
            ident,
            tok(Token::Colon),
            separated_list0(tok(Token::Semicolon), insn),
            tok(Token::Semicolon),
        )),
        |(label, _, insns, _)| BasicBlock {
            label: label.into_ident().unwrap(),
            insns,
        },
    )(i)
}

/// Parse a function body.
fn body(i: &[Token]) -> IResult<&[Token], Vec<BasicBlock>> {
    map(many0(basic_block), |blocks| blocks)(i)
}

/// Parse a function definition.
fn function(i: &[Token]) -> IResult<&[Token], Function> {
    map(
        tuple((signature, tok(LCb), body, tok(RCb))),
        |((return_type, name, args), _, blocks, _)| Function {
            return_type,
            args,
            name,
            blocks,
        },
    )(i)
}

/// Parse a module definition.
fn module(i: &[Token]) -> IResult<&[Token], Ast> {
    map(many0(function), |funs| Ast { funs })(i)
}

fn run_parser<'a, O, E: std::fmt::Debug>(
    mut parser: impl nom::Parser<&'a [Token], O, E>,
    tokens: &'a [Token],
) -> Result<O> {
    let (rest, o) = parser.parse(tokens).expect("no unrecover errors");
    if rest.is_empty() {
        Ok(o)
    } else {
        Err(Error::ParseError)
    }
}

/// Helps to parse input to the AST.
pub struct Parser<'a> {
    input: &'a str,
}

impl<'a> Parser<'a> {
    /// Makes a new [`Parser`].
    pub fn new(input: &'a str) -> Self {
        Self { input }
    }

    /// Parse an input to AST consuming parser.
    pub fn parse(self) -> Result<Ast> {
        let tokens = Lexer::new(self.input).lex()?;
        println!("{:?}", tokens);
        run_parser(module, &tokens)
    }
}

#[cfg(test)]
mod tests {
    use std::env;
    use std::io::Read;
    use std::path::PathBuf;

    use crate::lexer::{Keyword, Token};
    use crate::parser::run_parser;

    use super::super::{Error, Result};
    use super::{typ, Parser};

    const TEST_FILES_DIR_NAME: &str = "test_files";

    fn get_test_files_dir() -> PathBuf {
        PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap())
            .join("..")
            .join(TEST_FILES_DIR_NAME)
    }

    fn read_file(path: &PathBuf) -> Result<String> {
        (|| -> std::io::Result<String> {
            let mut content = String::with_capacity(64);
            let mut file = std::fs::File::open(path)?;
            file.read_to_string(&mut content)?;
            Ok(content)
        })()
        .map_err(Error::internal)
    }

    #[test]
    fn test_pointer_type() -> Result<()> {
        let tokens = [Token::Keyword(Keyword::Int), Token::Star, Token::Star];
        let typ = run_parser(typ, &tokens)?;
        assert_eq!(format!("{:?}", typ), "Ref(Ref(Int32))");
        Ok(())
    }

    #[test]
    fn test_simple_type() -> Result<()> {
        let tokens = [Token::Keyword(Keyword::Void)];
        let typ = run_parser(typ, &tokens)?;
        assert_eq!(format!("{:?}", typ), "Void");
        Ok(())
    }

    #[test]
    fn run_test_files() -> Result<()> {
        let test_dir = get_test_files_dir();

        for entry in std::fs::read_dir(test_dir).map_err(Error::internal)? {
            let entry = entry.map_err(Error::internal)?;
            let path = entry.path();

            if path.extension().and_then(|ext| ext.to_str()) == Some("ir") {
                println!("testing file: {:?}", path.file_name().unwrap());

                let mut expected_path = path.clone();
                expected_path.set_extension("ans");

                let expected = read_file(&expected_path)?;
                let prog = read_file(&path)?;

                let actual = format!("{}", Parser::new(&prog).parse()?);
                println!("{}\n", actual);

                assert_eq!(expected, actual);
                println!("\n")
            }
        }

        Ok(())
    }
}
